### DISCUSSION TITLE
Light bleed compensation only for first x layers of overhangs

### DISCUSSION THREAD

**Discussion by murk-sy** [sn4k3/UVtools#861] [Q&A] [answered]
I'm currently working with transparent Ultracur ST45 on SL1S. The material is great and prints nicely, but the light bleed is making Z accuracy suffer a lot, especially as I'm trying to print M8 threads. Figured I need to take up some fancier tools.

From my testing, I've determined the rough light bleed at the safe layer time.
For **consecutive** 0.10mm layers:
- 1 layer -> 0.20mm total bleed
- 2 layers -> 0.40mm total bleed
- 3 layers -> 0.60mm total bleed
- 4+ layers -> 0.80mm total bleed (maximum)

Since I have no x/y accuracy issues, I would prefer if curing is done at 100% when bleed is not an issue, to ensure the part comes out as it should. However, the current light bleed functionality would also dim the "core" parts of the print, which I would rather avoid if possible.

Examples (not to scale, illustrative, pink is bleed)
![image](https://github.com/sn4k3/UVtools/assets/1211672/482056b6-aac2-45f1-9057-c2589732a3ca)
![image](https://github.com/sn4k3/UVtools/assets/1211672/9f800c72-2372-4d9f-b570-ec872865be84)

I assume the standard UVtools (i.e. "what I managed") method *should* work perfectly fine once tuned, but I want to ensure that stuff that *can* be cured at 100% brightness actually *is*.

Am I overcomplicating things?

**Comment by sn4k3** [comment]
I can improve the tool to apply only to walls in the future.
You can however do what you want using a chain of tools runs.

1. Do up to "What I managed" step
2. Download and unzip this file [PixelArithmetic.zip](https://github.com/sn4k3/UVtools/files/14887192/PixelArithmetic.zip)
3. Drag and drop the "PixelArithmetic.uvtop" file into UVtools windows
4. Run the tool as it is
5. Report if is what you want

**Comment by murk-sy** [comment]
Thanks for the quick response. It's not quite what I'm looking for though, since pixelarithmetic doesn't account for what's above/below - unless I'm doing something wrong. Mspainted mockup:

![image](https://github.com/sn4k3/UVtools/assets/1211672/51240fae-e830-4678-9528-0795366b3f53)



Ended up making more examples, a truth table, and pseudocode to even wrap my own head around it.


Ideally this system should allow dimming to 0 and a custom amount of levels.
In examples below, yellow is slightly dimmed and red is most dimmed.

![image](https://github.com/sn4k3/UVtools/assets/1211672/91676735-ccdb-4793-b300-70313ab0082d) 

![image](https://github.com/sn4k3/UVtools/assets/1211672/74756255-a2b4-457e-b16a-c8d094d19c16) 
This I believe would be optimal for what I'm looking for and reduce possible stray pieces in the resin vat/on the FEP


![image](https://github.com/sn4k3/UVtools/assets/1211672/784f37d5-693a-42e1-9d32-09122e13cc55)
This would make it slightly riskier, but would likely achieve most accurate results for most technical parts at least.



Truth table and pseudocode to hopefully make it less confusing
![image](https://github.com/sn4k3/UVtools/assets/1211672/509b59e6-af2a-4808-8fa8-c3a28bfa603e)

```php
// assume 3 levels of dimness
// this code assumes a 1-dimensional array of single pixels, from z 0 to z max

foreach (pixelarray as pixel) {

	if (pixel is black) {
		pixel = black //do nothing
	}
	else if (pixel is white) { //or rather "non 0" to account for antialiasing
		if (layers -3, -2, -1 are white) {
			pixel = white //do nothing
		}
		
		if (layers +1, +2, +3 are white || option "minimise light bleed" is selected) {
			if (layers -3, -2, -1 are white) {
				pixel = white //do nothing
			}
			if (layers -3 are black && layers -2, -1 are white) {
				pixel = dim_1 // dim min
			}
			if (layers -2 are black && layers -1 are white) {
				pixel = dim_2 // dim medium
			}
			if (layers -1 are black ) {
				pixel = dim_3 // dim max
			}
		}
		
		
		if (layers +1, +2 are white && layers +3 are black) {
			if (layers -2, -1 are white) {
				pixel = white //do nothing
			}
			if (layers -2 are black && layers -1 are white) {
				pixel = dim_1 // dim min
			}
			if (layers -1 are black) {
				pixel = dim_2 // dim medium
			}
		}
				
		
		if (layers +1 are white && layers +2 are black) {
			if (layers -1 are white) {
				pixel = white //do nothing
			}
			if (layers -1 are black) {
				pixel = dim_1 // dim min
			}
		}
		
		
		if (layers +1 are black) {
			pixel = white //do nothing
		}
	}
}
```

**Reply by sn4k3** [reply, marked as answer]
That is not possible with current tools. Maybe you can take your pseudocode and try to make a script for UVtools with that logic and iterate up to the desired result.

**Comment by murk-sy** [comment]
Been a while since I did any C#, but I'll give it a try. If I end up with something workable I'll definitely share it.

**Reply by sn4k3** [reply]
You can see here some samples and structure of a script: https://github.com/sn4k3/UVtools/tree/master/Scripts/UVtools.ScriptSample
Also the code of light bleed: https://github.com/sn4k3/UVtools/blob/master/UVtools.Core/Operations/OperationLightBleedCompensation.cs#L200
